feat: modify user endpoint

rebelchris opened the pull request:
With regards to the frontend changes we no longer need to query the Gateway for user info.
This PR ensures we fetch users from the API directly (leveraging the new graphQL for it)

WT-298

Related to: https://github.com/dailydotdev/apps/pull/1141

idoshamun commented on src/schema/users.ts:
```diff
@@ -154,9 +160,13 @@ export const typeDefs = /* GraphQL */ `
     """
     infoConfirmed: Boolean
```
I know that it's not part of this PR and I might have missed that, but this should be provided only to the same logged-in user. For example, if Lee is trying to get Chris's `infoConfirmed` field it should be null for him. Only if Chris tries to get Chris's `infoConfirmed` it should have the right value

idoshamun commented on src/schema/users.ts:
```diff
@@ -491,7 +501,26 @@ export const resolvers: IResolvers<any, Context> = {
+        return builder;
+      });
+      if (!res[0]) {
+        throw new ForbiddenError('user not found');
```
Do we throw Forbidden or not found usually?

rebelchris replied:
We changed it, this is the general user call.
In the graphql adapter we made it forbidden error for the whoami call.

idoshamun replied:
In the whoami it makes sense because you are not logged in so you should receive forbidden if there's an issue.
But here the user doesn't exist, so it's better I think to throw NotFound

idoshamun commented on src/common/users.ts:
```diff
@@ -45,17 +45,6 @@ export const fetchUser = async (userId: string): Promise<User | null> => {
-  return res.json();
-};
-
 export const fetchUserInfo = async (userId: string): Promise<UserInfo> => {
```
`fetchUserInfo` is used mostly in the worker and needs to be updated.
`fetchUserRoles` I'm not sure where we used it, but it's worth checking.

rebelchris replied:
Alright, let me work on those in this PR

idoshamun reviewed (commented):
I think we should expand the scope of this PR, please see my comments.
What you implemented so far looks great

rebelchris commented:
@idoshamun Sorry, became quite big because the wide range of test cases that mocked the Gateway 😅
Addressed all issues mentioned.

As stated on Slack, we should consider making some fields required on a DB level at once stage.

idoshamun commented on src/common/users.ts:
```diff
@@ -34,42 +29,12 @@ const authorizedHeaders = (userId: string): { [key: string]: string } => ({
+  userId: string,
+  con: Connection,
+): Promise<User | null> => {
+  const user = await con.getRepository(DbUser).findOneOrFail({ id: userId });
```
I see that if the user doesn't exist we simply return null not failing

idoshamun commented on src/entity/User.ts:
```diff
@@ -88,6 +89,14 @@ export class User {
+
+  permalink: string;
+  @AfterLoad()
+  setComputed() {
```
Nice!!

idoshamun commented on src/entity/User.ts:
```diff
@@ -18,13 +19,13 @@ export class User {
 
   @Column({ type: 'text', nullable: true })
-  name?: string;
+  name: string;
```
We need to remember to generate a migration file afterwards

idoshamun commented on src/workers/communityLinkAccessMail.ts:
```diff
@@ -25,6 +25,7 @@ const worker: Worker = {
           data.userId,
       );
     } catch (err) {
+      console.log(err);
```
Not needed

idoshamun reviewed (approved):
Some comments, overall looks great!
